src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,17 +105,170 @@ def pileupSizeTask(self):
+        """
+        Execute partial pileup placement according to the following logic:
+
+        - get rucio DIDs (datasets, in CMS name blocks) for our pileup document
+        - get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+        - create new container DID in Rucio via Rucio wrapper createContainer API
+          which calls rucio CLI add_container
+        - update MSPileup JSON with
+          - create new custom Name as pileup+extention
+          - add new transition record and update MSPileup document
+        - we call attachDIDs Rucio wrapper API with our set of DIDs and rses from pileup document
+        - create new rules for custom DID via provided rseExpression
+        - add new ruleIds to pileup document
+        - set expiration date (to be 24h) for already existing ruleIds from pileup document
+
+        :return: None
+        """
+        spec = {'active': True}
+        docs = self.mgr.getPileup(spec)
+        for doc in docs:
+            self.logger.info("partialPileupTask process %s", doc)
+            fraction = doc['containerFraction']
+
+            # check previous fraction value and proceed only if has changed
+            previousFraction = 1.0
+            transition = doc.get('transition', [])
+            if transition:
+                previousFraction = transition[-1]['containerFraction']
+            if previousFraction == fraction:
+                # do nothing since there is no transition, i.e. skip this record
+                self.logger.info("Pileup name %s has no container fractions changes", doc['pileupName'])
+                continue
+
+            # get rucio DIDs (datasets, in CMS name blocks) for our pileup document as following:
+            # - when container fraction is decreasing: new container is a subset of customName
+            # - when container fraction is increasing: new container is a superset of customName + subset of pileupName
+
+            # usage of block names defined in this logic:
+            # https://github.com/dmwm/WMCore/pull/11807#pullrequestreview-1786778783
+            customBlocks = self.getPileupBlocks(doc, previousFraction, fraction)
+
+            # create new container DID in Rucio for our custom name
+            self.rucioClient.createContainer(doc['customName'], scope=self.customRucioScope)
+            self.logger.info("Create container %s with scope %s", doc['customName'], self.customRucioScope)
+
+            # call rucio APIs to attach custom blocks to our custom container (DID)
+            newRules = []
+            self.rucioClient.attachDIDs(None, doc['customName'], customBlocks, scope=self.customRucioScope)
+            for rse in doc['expectedRSEs']:
+                # create new rule for custom DID using pileup document rse
+                newRules += self.rucioClient.createReplicationRule(doc['customName'], rse)
+
+            self.logger.info("Custom pileup: %s has the following new rules created: %s for RSEs: %s",
```

Comment:
Given that a problem can happen at any time, including one rule creation succeeding and not the others, I would refactor this log record and move it to the line above AND inside the for loop. Message would be something like "Rule id: %s created for custom pileup: %s for RSE: %s"

Reply:
ok